File: Leetcode/Odd Even Linked List/Odd Even Linked List.py
class Solution:
    def oddEvenList(self, head):
        if not head:
            return head
        
        odd = head # Both of them point at the first node of the target linked list
        even = head.next # doesn't matter even there's only one node in the linked list (even will become None)
        eHead = even # have to keep where the even-node list starts
        
        while even and even.next: # won't get in the loop at first if there's only one node in the linked list
            # both even and even.next are necessary condition because even might point to None, which has no attribute 'next'
            # AND, why these two, small discussion by myself as below
            odd.next = odd.next.next
            even.next = even.next.next
            # After these two ops, odd/even still points at its original place
            # Therefore, move them to the next node repectively
            odd = odd.next
            even = even.next
        
        odd.next = eHead # the odd pointer currently points at the last node of the odd-node list
        
        return head # keep the start of the odd-node list in the first of our code

File: Leetcode/Odd Even Linked List/Odd Even Linked List.c
//Problem:https://leetcode.com/problems/odd-even-linked-list/

struct ListNode* oddEvenList(struct ListNode* head) {
    // create a start node 
    struct ListNode* start = malloc(sizeof(struct ListNode));
    start->next = head;
    
    struct ListNode* odd = start, * curpre = start, * cur = head;
    bool isodd = true;
    while (cur != NULL) {
        if (isodd && odd->next != cur) {
            // if cur is odd, cur should be moved to odd's next
            curpre->next = cur->next;
            cur->next = odd->next;
            odd->next = cur;
            cur = curpre->next;
            odd = odd->next;
        } else {
            // if cur is odd but odd's next is already pointed to cur, 
            // nothing will be changed, just move the odd pointer
            odd = isodd ? odd->next : odd;
            curpre = cur;
            cur = cur->next;
        }
        isodd = !isodd;
    }
    return start->next;
}
